### ISSUE TITLE
Potential for allowing adding devices by IP

### ISSUE THREAD

**Issue by laf** [librenms/librenms#594]
Two ways we could do this:

1. Drop the requirement for hostnames in full and just allow IPs. If we did this then we would need to automatically disable xDP discovery as this would cause devices to be added again if hostnames actually did exist and just weren't used.

2. Areas that would need to be updated to support this:

 - New flag in devices table to acknowledge this is a manual hostname
 - New column in devices for IP address (ipv6 as well?)
 - Add device in webui needs updating.
 - Add device in shell.
 - API to add device.

Using this new flag:

 - Disable xDP.
 - Change from using hostname to IP.

What about the alerts?

**Comment by laf** [comment]
Tagging @paulgear and @f0o for thoughts.

**Comment by paulgear** [comment]
I can't help but think that the main reason this wasn't done before is because it's too hard to handle all the corner cases.  I'm happy to see it added in, but I would consider it low priority, especially when anyone can put in place whatever workarounds they like in /etc/hosts.

**Comment by laf** [comment]
I'm 'fairly' certain it's a bit of a hangup from adam about enforcing a strict way to manage devices and xDP potentially causing issues.

Maybe we could add a config option that's set to disable the feature by default but it can be turned on - with warning?

I might just remove the check for hostname on my setup and see what doesn't work....

**Comment by f0o** [comment]
Assuming:
- user enters Name and IPv4/6
- there is a boolean flag in the devices-table that sais the hostname is non-DNS

I see no issues with alerting. XDP would require a small check on the ports table if there is a device associated with the discovered IP.

Or am I mistaken?

**Comment by laf** [comment]
To find out if the device discovered has an IP that already exists then you would only know this once a discovery is done so it's too late then. You could obviously check if the name returned from say cdp resolves to the IP but this may not be the case.

It would much safer to disable xDP for IP based devices, you can still run xDP for devices with hostnames.

If I had to vote on which option to go for I'd possibly say number 1 with the config option needing to be enabled to support it.

**Comment by f0o** [comment]
Re: autodiscovery; all we need to do is to patch a check in https://github.com/librenms/librenms/blob/cb9d3d1cbbe2c1f29f69c5831b23a3dde4656c17/includes/discovery/functions.inc.php#L14 to make sure there is no host with that IP known to librenms in order to avoid dupes when an existing fqdn wasnt used.

This shouldnt be so difficult. So overall xDP should work regardless of the presence of a dns-resolvable fqdn.. Dont you think?

**Comment by laf** [comment]
So long as it doesn't affect auto discovery in the way it works now then we should be fine. The IPs in ipv4_addresses don't have to be unique though so you can have the same IP duplicated twice so a check solely against that could result in false positives - I think

**Comment by wzhsunn** [comment]
I saw your talking about allowing [adding devices by ip](https://groups.google.com/forum/#!topic/librenms-project/ysTQaWrka5I) on google groups . I need this feature. I am looking forward to it . Thank you all .

**Comment by laf** [comment]
@paulgear Any further view on if you'd like someone to tackle us enabling this or are we continuing with hostname support only?

**Comment by paulgear** [comment]
@laf If people want it, and you're happy to work on it, I say go for it.  My only request would be to optimise for the hostname present case rather than the hostname absent case.

**Comment by laf** [comment]
For those wanting to give this a try, we have a separate branch that you can switch to and test.

BE WARNED - WE CANNOT SAY THIS WON'T BREAK SOMETHING (but it really shouldn't :))

If you want to use it, in your LibreNMS install do the following:

git remote update
git checkout ip-devices

Now play around, add devices, do what you want to give it a whirl and report issues in here.

Want to go back to the master branch, easy:

git checkout master
git pull

Please be aware though that you won't receive any updates whilst you are on the ip-devices branch, if you switch back to master, I'd suggest you delete any devices added by IP only.

**Comment by laf** [comment]
Anyone tested this out and can feedback?

**Comment by tylerapplebaum** [comment]
Works! Thank you SO much for this.

**Comment by laf** [comment]
Are you giving it a test as we'd be interested in feedback. Especially if you use any auto discovery.

**Comment by laf** [comment]
Folks, I'll be working on having some validation code added so we don't add devices with the same IP and then we will look at releasing this.

**Comment by f0o** [comment]
:+1:

**Comment by laf** [comment]
A simple check for IP duplication is in PR #1248 so closing this now.

**Comment by Snpbond** [comment]
Did this functionality remain in current builds? CDP discovery is failing for me and it appears to be because it's attempting to discover by hostname and not IP.

```
discovering agnes-server-2960-4
name lookup of agnes-server-2960-4 failed
696, cdp, 0, agnes-server-2960-4, GigabitEthernet0/1, cisco WS-C2960-24TT-L, Cisco IOS Software, C2960 Software (C2960-
LANBASE-M), Version 12.2(35)SE5, RELEASE SOFTWARE (fc1)
```

I found on Line 58 of librenms/includes/discovery/discovery-protocols.inc.php it's attempting to add by cdpCacheDeviceId

Here is my CDP Array

```
(
            [cdpCacheAddressType] => ip
            [cdpCacheAddress] => 0A C1 D4 F9 
            [cdpCacheVersion] => Cisco IOS Software, C2960 Software (C2960-LANBASE-M), Version 12.2(35)SE5, RELEASE SOF
TWARE (fc1)
            [cdpCacheDeviceId] => agnes-server-2960-4
            [cdpCacheDevicePort] => GigabitEthernet0/1
            [cdpCachePlatform] => cisco WS-C2960-24TT-L
            [cdpCacheCapabilities] => 00 00 00 28 
            [cdpCacheVTPMgmtDomain] => 
            [cdpCacheNativeVLAN] => 1
            [cdpCacheDuplex] => fullduplex
            [cdpCacheLastChange] => 461:14:13:56.04
        )
```

The cdpCacheAddress converted from Hex is the switch IP, if I run the discover_new_device function with that converted IP then it seems to start crawling CDP Neighbors correctly.

**Comment by laf** [comment]
We don't do anything with CacheAddress, I'd suggest creating a new issue with this request in it.

**Comment by Snpbond** [comment]
Will do! Thanks
